viewing_party/party.py

```diff
@@ -1,23 +1,320 @@
+    """
+    total_rating = 0
+    movies = 0
+    for movie in user_data["watched"]:
+        movies += 1
+        total_rating += movie["rating"]
+        
+    if movies > 0:    
+        average_rating = total_rating/movies   
+    else:
+        average_rating = 0.0   
+
+    return average_rating
+
+def get_most_watched_genre(user_data):
+    """
+    Get the most frequently watched genre from user's watched movies
+    
+    Parameter:
+        user_data (dict): 
+            the key "watched" is a list of dictionaries, where each dictionary represent a movie
+            each movie dict contains a key "genre" (str)
+    Returns:
+        str: the genre that user watched frequently
+        None: if "watched" list is empty
+    """
+    genres_count = {} # map from genre to count
+
+    # Return None if there is no value in "watched."
+    if user_data["watched"] == []:
+        return None
+    
+    # Add and count the genres found. 
+    for movie in user_data["watched"]:
+        genre = movie["genre"]
+        if genre not in genres_count:
+            genres_count[genre] = 1
+        else:
+            genres_count[genre] += 1
+            
+    max_rate = 0
+    most_frequent_genre = None
+    # Comparing the rates and returning the most frequenty genre.
+    for key, rate in genres_count.items():
+        if rate > max_rate:
+            most_frequent_genre = key
+            max_rate = rate    
+    
+    # using max() to get the max count in genere_count dictionary       
+    # most_frequent_genre = max(genres_count, key=genres_count.get) 
+    # If use max(dictionary), max() looks only at the keys of the dictionary, not the values.
+	# If the keys are strings, max() will return the key that comes last alphabetically   
+               
+    return most_frequent_genre
 
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
+def get_friends_watched(user_data):
+    """
+    Helper function to get list of all movies that friends have watched
+    """
+    friends_watched= []
+    
+    for friend in user_data["friends"]:
+        friends_watched.extend(friend["watched"])
         
+    return friends_watched
+
+
+def get_unique_watched(user_data):
+    """
+    This function determines which movies the user has watched 
+    but none of the friends have watched.
+    
+    Parameters:
+        user_data(dict): has two keys:
+            "watched": List of movie dictionaries the user has watched, each with a "title".
+            "friends": List of friends, each having a "watched" list of movie dictionaries.
+
+    Returns:
+        list of dict of movies.
+    """
+    
+    user_watched = user_data["watched"]
+    friends_watched = get_friends_watched(user_data)
+    
+    unique_watched = []
+
+    for movie in user_watched:
+        if movie not in friends_watched:
+            unique_watched.append(movie)
+
+    return unique_watched

```

Comment:
Great use of helper functions here! This is a really good example of abstracting out certain details to make your life easier!